docs/proposals-accepted/202205-vertical-query-sharding.md

```diff
@@ -0,0 +1,152 @@
+
+### Pitfalls of the current solution
+
+When executing a PromQL query, a Querier will pull series from all of its downstream Stores in memory and feed them to the Prometheus query engine. In cases where a query causes a large number of series to be fetched, executing the query can lead to high memory usage in Queriers. As a result, even when more than one Querier is available, work is concentrated in a single Querier and cannot be easily distributed.
+
+## Goals
+* **Must** allow the sharding of aggregation expressions across N queries, where N denotes a user-provided sharding factor  
+* **Must** fall back to single query when a non-shardable query or querier is encountered in the query path
+* **Could** serve as a motivation for a general query-sharding framework that can be applied to any query 
+* **Should not** make any changes to the current query request path (i.e. how queries are scheduled in Queriers)
+
+### Audience
+
+* Users who run Thanos at a large scale and would like to benefit from improved stability of the Thanos read path.
+
+## How
+
+### The query sharding algorithm
+
+The query sharding algorithm takes advantage of the grouping labels provided in PromQL expressions and can be applied to the large majority of PromQL queries which aggregate timeseries _by_ or _without_ one more grouping labels. 
+
+In order to illustrate how it works, we can take the example of executing the following PromQL query
+```
+sum by (pod) (memory_usage_bytes)
+```
+on the series set:
+ 
+```
+memory_usage_bytes{pod="prometheus-1", region="eu-1", role="apps"}
+memory_usage_bytes{pod="prometheus-1", region="us-1", role="infra"}
+memory_usage_bytes{pod="prometheus-2", region="eu-1", role="apps"}
+memory_usage_bytes{pod="prometheus-2", region="us-1", role="infra"}
+```
+
+This query performs a sum over series for each individual pod, and is equivalent to the union of following two queries:
+
+```
+sum by (pod) (memory_usage_bytes{pod="prometheus-1"})
+```
+and 
+```
+sum by (pod) (memory_usage_bytes{pod="prometheus-2"})
+```
+
+We can therefore execute each one of the two queries in a separate query shard and concatenate the result before returning it to the user.
+
+### Dynamic series partitioning
+
+Since Queriers have no information about which timeseries are available in Stores, they cannot easily rewrite a single aggregation into two disjoint ones. They can, however, propagate information to Stores that instruct them to only return one disjoint shard of the series that match particular selectors. To achieve this, each query shard would propagate the total number of shards, its own shard index, and the grouping labels discovered in the PromQL expression. Stores would then perform a `hashmod` on the grouping labels against each series and return only those series whose `hashmod` equals the requested shard index.
```

Comment:
Mimir's split and merge compactor shards a large block to multiple small blocks. Each small-block contains all series for a specific shard.
If the store gateway is shard-aware or each small block contains a special label indicating the shard id, then we don't have to do `hashmod` during query times and performance can be improved since less blocks need to be queried.

I am not sure if we can reuse that split and merge compactor, but we can do something similar for compaction time sharding.

Reply:
This is indeed something we considered, but in the end we were concerned with the complexity of the approach. The way I understand this, we either have to implement some sort of sharding at write, which would likely not work for the Prometheus sidecar, or make users configure disjoint shards themselves, which could break backwards compatibility. 

It also turns out that calculating `hashmod` is quite cheap and can be done on the fly. The only overhead is the multiple requests that go to the Stores, but for receivers and store gateways this is not a problem because sharding is done before chunks are even retrieved. With the Prometheus sidecar it could be a concern because sharding is done once series with chunks are retrieved, but for the time being there isn't really a good way to solve this. 

If sharding is added to the TSDB, we can modify our approach to take advantage of that, but we thought it was more important to start with something that already delivers value.

Reply:
I just tried a high cardinality query against Prometheus with and without sharding. So with sharding, retrieving series took 1.7 seconds (hashing was just 0.01s) and the overall query took 3.5 seconds. Without sharding, retrieval took 1.3 seconds but the entire query took 6.5 seconds.

Reply:
Thanks. I think the current one is good as a start and we can iterate.
As the tsdb shard index has not been upstreamed yet, we can think about it later.

Reply:
My 2 cents: I think sharding could happen in many places simultaneously, e.g. just like we shard requested cache keys into batches and so on. The idea, in general, is the same - to perform more work at the same time i.e. use resources more efficiently. It could be trickier to implement fair resource sharing between tenants in the future with sharding in many places. Still, I think it's a non-problem currently since usually multi-tenancy is achieved by each tenant having their own read/write paths with something like shuffle sharding i.e. separate processes in general.

Reply:
Can we also mention the compaction time sharding for store gateway in this proposal? We don't have to implement this for phase 1 but this is something nice to have.

Reply:
Should we mention this under alternatives?

Reply:
It is a good complement to the current proposal. I think it is not really an alternative and we can have both.

Reply:
I added it under a `Future improvements` section